### PR TITLE
gh-99880: document rounding mode for new-style formatting

### PR DESCRIPTION
**Pull request by skirpichev**
The CPython uses _Py_dg_dtoa(), which does rounding to nearest with half to even tie-breaking rule.

If that functions is unavailable, PyOS_double_to_string() fallbacks to system snprintf().  Since CPython 3.12, build requirements include C11 compiler *and* support for IEEE 754 floating-point numbers. C17 says here (p 230, "Recommended practice"): "For e, E, f, F, g, and G conversions, if the number of significant decimal digits is at most DECIMAL_DIG, then the result should be correctly rounded."

* Issue: gh-99880

----
📚 Documentation preview 📚: https://cpython-previews--121481.org.readthedocs.build/

### CODE REVIEW

**Comment by jowagner** [comment]
My opinion (not a core developer):
- The word "correctly" doesn't add anything useful here as correctness is subjective, i.e. what is the correct way of rounding for one person may be wrong to another. Suggest removing this word.
- Not clear what kind of vagueness is expressed with "it should round" here. Suggestion: "it will, on common platforms, round".
- Many readers may not understand what least significant bits have to do with decimals. Suggestion: "[...] toward the value with an even last digit (0, 2, 4, 6 or 8) in its representation." (There are always two nearest values and one is always odd and the other even, so there is exactly one even value. Note also that "even" and "odd" are not properties of numbers alone but only of numbers with a selected format, e.g. 3 is odd for `.0f` but even for `.1f`.)

File: Doc/library/string.rst [hunk 1]
```diff
@@ -588,6 +588,13 @@ The available presentation types for :class:`float` and
    |         | as altered by the other format modifiers.                |
    +---------+----------------------------------------------------------+
 
+The result should be correctly rounded to a given precision ``p`` of digits
+after the decimal point.  For :class:`float`, used rounding mode match the one
+for :func:`round` builtin, i.e. it should round results midway between
```
Comment (skirpichev) [inline Doc/library/string.rst]:
```suggestion
for :func:`round` builtin, i.e. it will round results midway between
```
?

**Review by skirpichev — COMMENTED** [review]
@jowagner, thanks for review.

> The word "correctly" doesn't add anything useful here as correctness is subjective

The meaning in this context actually is not subjective at all.  From § 3.9 of the C11 standard: correctly rounded result - representation in the result format that is nearest in value, subject to the current rounding mode, to what the result would be given unlimited range and precision.

Maybe it worth to add something like above definition to the glossary?  It's used a lot e.g. in the decimal module.

> Suggestion: "it will, on common platforms, round".

+1

> Many readers may not understand what least significant bits have to do with decimals.

Well, it's a *bit* not a *digit*;)  Hardly we should everywhere emphasize that floats are *binary* floating point numbers.  I would prefer if more detailed description (in depth, with examples and with a warning!) will be consolidated in round() docs.  There is a link.

> Suggestion: "[...] toward the value with an even last digit (0, 2, 4, 6 or 8) in its representation." 

I doubt it's a correct description.  Consider this example:
```pycon
>>> round(2.675, 2)
2.67
>>> from decimal import Decimal, getcontext
>>> getcontext().rounding
'ROUND_HALF_EVEN'
>>> round(Decimal('2.675'), 2)
Decimal('2.68')
```

File: Doc/library/string.rst [hunk 2]
```diff
@@ -588,6 +588,13 @@ The available presentation types for :class:`float` and
...
+representable values toward the nearest value with an even (0) least
+significant bit.  For :class:`~decimal.Decimal`, rounding mode of the current
+:ref:`context <decimal-context>` is used.
```
Comment (picnixz) [inline Doc/library/string.rst]:
```suggestion
The result should be correctly rounded to a given precision ``p`` of digits
after the decimal point.  The rounding mode for :class:`float` matches that
of the :func:`round` builtin.  For :class:`~decimal.Decimal`, the rounding
mode of the current :ref:`context <decimal-context>` will be used.
```

No need to restate what is already specified for `round`. People could just click on the corresponding link.

Comment (picnixz) [reply Doc/library/string.rst]: Ok, so it's not a ROUND_HALF_TO_EVEN. But I don't know how to call it. It's ROUND_HALF_TO_EVEN_NDIGITS instead but is there an official name for it?

Comment (skirpichev) [reply Doc/library/string.rst]: > Ok, so it's not a ROUND_HALF_TO_EVEN.

It's ``FE_TONEAREST`` of Annex F or ``ROUND_HALF_EVEN`` in the decimal module.

**Review by picnixz — APPROVED** [review]

**Review by willingc — COMMENTED** [review]
This looks good to me @skirpichev. There is a merge conflict that needs resolution. Thanks.

**Comment by skirpichev** [comment]
Thanks, fixed.

**Review by willingc — APPROVED** [review]
Thanks @skirpichev.

**Pull request merged**